### PR TITLE
Add EXAONE 4.0 model support for Inference V2

### LINKED ISSUE #7453: [REQUEST] Add support for EXAONE 4.0 models
Hi DeepSpeed team,

I’d like to suggest adding support for **EXAONE 4.0** (LGAI‑EXAONE/EXAONE‑4.0‑32B and 1.2B) in DeepSpeed.  
Currently, this model is officially supported by vLLM (via PR #21060), but DeepSpeed doesn’t yet recognize or integrate it.  
This means we can’t fully take advantage of ZeRO, offload, and tensor parallel features with EXAONE 4.0.

EXAONE 4.0 is a hybrid LLM (combining reasoning and non‑reasoning modes) and provides strong performance on reasoning tasks and multilingual content (Korean, English, Spanish). Benchmarks like MMLU‑Redux and MATH show impressive numbers (MMLU‑Redux 92.3, MATH/AIME ~85%).

**What we’d like to see:**
1. Register `Exaone4ForCausalLM` in DeepSpeed’s supported model registry.
2. Add support for initialization, parameter loading, and optimizer state management for the EXAONE 4.0 architecture.
3. Ensure compatibility with ZeRO-offload (CPU/NVMe) and tensor parallel inference.

I’d be happy to share test logs or work on a draft PR if needed.  
Adding EXAONE 4.0 support would be a great benefit for the growing Korean and multilingual AI community.

Thanks for your time and consideration!

### PR DESCRIPTION
**Pull request by Bias92**
## Summary
Add support for LG AI Research's EXAONE 4.0 model family in DeepSpeed Inference V2.

Closes #7453

## Changes
- New model implementation: `deepspeed/inference/v2/model_implementations/exaone4/`
  - `container.py`: Transformer and non-transformer parameter containers
  - `model.py`: Inference model with post-norm architecture and QK-Norm support
  - `policy.py`: Inference V2 policy
- Register EXAONE 4.0 in `engine_factory.py` and `__init__.py`

## Key architectural differences from Mistral/Llama
- **Post-norm**: RMSNorm is applied after attention/MLP outputs (not before), followed by residual addition
- **QK-Norm**: Per-head RMSNorm applied to Q and K projections after the QKV linear layer
- **Hybrid attention**: 32B model uses 3:1 sliding window/full attention ratio (via `layer_types` config)

## Supported models
- [EXAONE-4.0-1.2B](https://huggingface.co/LGAI-EXAONE/EXAONE-4.0-1.2B) (all full attention)
- [EXAONE-4.0-32B](https://huggingface.co/LGAI-EXAONE/EXAONE-4.0-32B) (hybrid sliding/full attention)

Requires `transformers >= 4.54.0`.

## Related
- Supersedes #7456 (draft, inactive for 6 months)

### CODE REVIEW

File: deepspeed/inference/v2/model_implementations/exaone4/model.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+    def _apply_qk_norm(self, hidden_states: torch.Tensor, q_norm_gamma: torch.Tensor,
+                       k_norm_gamma: torch.Tensor) -> torch.Tensor:
+        """
+        Apply RMSNorm to Q and K projections independently per head.
+        hidden_states shape: [tokens, (n_q + n_kv + n_kv) * head_size]
+        """
+        tokens = hidden_states.shape[0]
+        local_n_heads = self.n_heads // max(self.tp_size, 1)
+        local_n_heads_kv = self.n_heads_kv // max(self.tp_size, 1)
```
Comment (tohtana) [inline deepspeed/inference/v2/model_implementations/exaone4/model.py]: As EXAONE4 has uneven Q/KV heads (GQA), I think this can produce incorrect results. Shouldn't we use these?
- `self.n_heads_q_local` instead of `self.n_heads // self.tp_size`
- `self.n_heads_kv_local` instead of `self.n_heads_kv // self.tp_size`

**Review by tohtana — COMMENTED** [review]
@Bias92 Thank you for your contribution!
I left a comment about QKV partitioning. Can you check it?
I would be great if you could validate that the model implementation produces coherent text.

**Comment by Bias92** [comment]
@tohtana Thanks for the review! I've updated the code to use `self.n_heads_q_local` and `self.n_heads_kv_local` as suggested.

I'll validate the model with coherent text generation and share the results.

**Comment by Bias92** [comment]
### Validation: coherent text generation with EXAONE-4.0-1.2B

**Environment**: RTX 4060 Ti, PyTorch 2.10, transformers 5.1.0, CUDA 12.8
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

model = AutoModelForCausalLM.from_pretrained("LGAI-EXAONE/EXAONE-4.0-1.2B", dtype=torch.float16, device_map="auto", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained("LGAI-EXAONE/EXAONE-4.0-1.2B", trust_remote_code=True)

prompt = '[|user|]\nExplain what DeepSpeed is in 2 sentences.\n[|assistant|]\n'
inputs = tokenizer(prompt, return_tensors="pt").to("cuda")
outputs = model.generate(**inputs, max_new_tokens=100, do_sample=False)
print(tokenizer.decode(outputs[0], skip_special_tokens=True))
```

The model produces coherent, contextually relevant text. Full end-to-end validation with DeepSpeed Inference V2 would require multi-GPU environment — happy to assist with further testing if needed. 
image

**Review by tohtana — APPROVED** [review]
@Bias92 Thank you for the update! This looks good to me.
